gen/lib/gen/generator.rb

```diff
@@ -0,0 +1,163 @@
+require 'gen'
+require 'fileutils'
+require 'open3'
+require 'pathname'
+require 'tmpdir'
+
+module Gen
+  class Generator
+    def self.run(project_name)
+      new(project_name).run
+    end
+
+    TEMPLATE_ROOT = File.expand_path('gen/template', Gen::ROOT)
+
+    VALID_PROJECT_NAME = /\A[a-z][a-z0-9]*\z/
+    private_constant :VALID_PROJECT_NAME
+
+    # false  -> delete file
+    # string -> rename file before applying template substitutions
+    VENDOR_TRANSLATIONS = {
+      'Gemfile'            => false,
+      'exe/__app__-gems'   => false,
+      'exe/__app__-vendor' => 'exe/__app__',
+      'dev-gems.yml'       => false,
+      'dev-vendor.yml'     => 'dev.yml',
+    }.freeze
+    private_constant :VENDOR_TRANSLATIONS
+
+    BUNDLER_TRANSLATIONS = {
+      'bin'                => false,
+      'bin/update-deps'    => false,
+      'exe/__app__-gems'   => 'exe/__app__',
+      'exe/__app__-vendor' => false,
+      'dev-gems.yml'       => 'dev.yml',
+      'dev-vendor.yml'     => false,
+    }.freeze
+    private_constant :BUNDLER_TRANSLATIONS
+
+    def initialize(project_name)
+      raise(
+        CLI::Kit::Abort,
+        "project name must match {{bold:#{VALID_PROJECT_NAME}}} (but can be changed later)"
+      ) unless project_name =~ VALID_PROJECT_NAME
+      @project_name = project_name
+      @title_case_project_name = @project_name.sub(/^./, &:upcase)
+    end
+
+    def run
+      vendor = ask_vendor?
+      create_project_dir
+      if vendor
+        copy_files(translations: VENDOR_TRANSLATIONS)
+        update_deps
+      else
+        copy_files(translations: BUNDLER_TRANSLATIONS)
+      end
+    end
+
+    private
+
+    def ask_vendor?
+      vendor = nil
+      CLI::UI::Frame.open('Configuration') do
+        q = 'How would you like the application to consume {{command:cli-kit}} and {{command:cli-ui}}?'
+        vendor = CLI::UI::Prompt.ask(q) do |c|
+          c.option('Vendor  {{italic:(faster execution, more difficult to update deps)}}') { 'vendor' }
+          c.option('Bundler {{italic:(slower execution, easier dep management)}}') { 'bundler' }
+        end
+      end
+      vendor == 'vendor'
+    end
+
+    def create_project_dir
+      info(create: '')
+      FileUtils.mkdir(@project_name)
+    rescue Errno::EEXIST
+      error("directory already exists: #{@project_name}")
+    end
+
+    def copy_files(translations:)
+      each_template_file do |source_name|
```

Comment:
I feel like we should yield `source_name` and `source_file`, then we don't need to have the File.expand_path below. This also keeps the source file logic in each_template_file and disassociates the constant use from here, which I feel encapsulates logic more effectively

Reply:
I tried this but IMO it looks better as written with the source / target assignments one after another.